File: client/src/models/Agent.ts
import type Admin from "./Admin";
import type Buy from "./Buy";
import type Transfer from "./Transfer";

type AgentJSON = {
  id?: string;
  name?: string;
  email?: string;
  status?: string;
  password?: string;
  admin_id?: string;
  balance?: number;
};
class Agent {
  #id: string;
  name: string;
  balance: number;
  status: string;
  email: string;
  #admin_ref: Admin | null;
  constructor(agentJSON: AgentJSON, admin?: Admin) {
    this.#id = agentJSON.id || "";
    this.#admin_ref = admin || null;
    this.balance = agentJSON.balance || 0;
    this.name = agentJSON.name || "";
    this.status = agentJSON.status ?? "active";
    this.email = agentJSON.email ?? "";
  }

  get id(): string {
    return this.#id;
  }

  get transfers(): Transfer[] {
    // return this.admin?.transfers || [];
    return this.admin?.get_transfer_by_agent(this.#id) || [];
  }

  get admin(): Admin | null {
    return this.#admin_ref;
  }
  get buys(): Buy[] {
    return this.#admin_ref?.buys.filter((b) => b.agent?.id === this.#id) || [];
  }
}

export type { AgentJSON };
export default Agent;

File: server/models/agent.go
package models

import (
	"fmt"

	"github.com/google/uuid"
)

type Agent struct {
	ID       uuid.UUID  `gorm:"column:id;type:uuid;primaryKey;default:uuid_generate_v4()" json:"id"`
	Name     string     `gorm:"column:name;not null" json:"name"`
	Balance  float64    `gorm:"column:balance;default:0" json:"balance"`
	AdminID  uuid.UUID  `gorm:"column:admin_id;type:uuid;index" json:"admin_id"`
	Admin    *Admin     `gorm:"foreignKey:AdminID;constraint:OnUpdate:CASCADE,OnDelete:SET NULL" json:"admin"`
	Transfer []Transfer `gorm:"foreignKey:AgentID" json:"transfers"`
	Buys     []Buy      `gorm:"foreignKey:AgentID"`
}

func (Agent) TableName() string {
	return "agent"
}

func (a Agent) String() string {
	smnt := ""
	smnt += fmt.Sprintf("Agent(id='%s', name='%s', balance=%f, admin_id=%s", a.ID, a.Name, a.Balance, a.AdminID)
	if a.Admin != nil {
		smnt += fmt.Sprintf(", admin='<Admin(%s)>'", a.Admin.Name)
	}
	smnt += ")"
	return smnt
}
